### ISSUE TITLE
BatchNormalization: validation loss >> train loss, same data

### ISSUE THREAD

**Issue by OverLordGoldDragon** [keras-team/keras#12851]
I implemented [this paper's][1] neural net, with some differences (img below), for EEG classification; ```train_on_batch``` performance is excellent, with very low loss - but ```test_on_batch``` performance, though on *same data*, is poor: the net seems to always predict '1', most of the time:

```
       TRAIN (loss,acc)   VAL (loss,acc)
'0' -- (0.06269842,1)     (3.7652588,0)
'1' -- (0.04473557,1)     (0.3251827,1)
```
Data is fed as 30-sec segments (12000 timesteps) (10 mins per dataset) from 32 (= batch_size) datasets at once (img below)

Any remedy?

<hr>
<b>Troubleshooting attempted</b>:<br>

   1. Disabling dropout
   2. Disabling all regularizers (except batch-norm)
   3. Randomly, val_acc('0','1') = (~.90, ~.12) - then back to (0,1)

<strong>Additional details</strong>:<br>

 - Keras 2.2.4 (TensorFlow backend), Python 3.6, Spyder 3.3.4 via Anaconda
 - CuDNN LSTM stateful<br>
 - CNNs pretrained, LSTMs added on afterwards (and both trained)
 - ```BatchNormalization``` after every CNN & LSTM layer<br>
 -  ```reset_states()``` applied between different datasets
 - [squeeze_excite_block][2] inserted after every but last CNN block

<hr>
<b>UPDATE</b>:
Progress was made; `batch_normalization` and `dropout` are the major culprits. Major changes:

 - Removed LSTMs, GaussianNoise, SqueezeExcite blocks (img below)
 - Implemented batch_norm [patch](https://github.com/keras-team/keras/pull/9965)
 - Added `sample_weights` to reflect class imbalance - varied between 0.75 and 2. 
 - Trained with various warmup schemes for both MaxPool and Input dropouts

Considerable improvements were observed - but not nearly total. Train vs. validation loss behavior is truly bizarre - flipping class predictions, and bombing the exact same datasets it had _just trained on_: 

<img src="https://user-images.githubusercontent.com/16495490/59074372-19754380-8899-11e9-8be4-4d9261ee880d.png" width="250">

Also, `BatchNormalization` outputs during train vs. test time differ considerably (img below)
<hr>
<b>UPDATE 2:</b> All other suspicions were ruled out: <b><u>BatchNormalization</u></b> is the culprit. Using Self-Normalizing Networks (SNNs) with SELU & AlphaDropout in place of BatchNormalization yields stable and consistent results.
<hr>
<img src="https://user-images.githubusercontent.com/16495490/59074862-8be72300-889b-11e9-8acc-bdc390d7466d.png" width="400">
<img src="https://i.stack.imgur.com/jp4fG.png" width="500">
<img src="https://user-images.githubusercontent.com/16495490/59076170-a8d32480-88a2-11e9-98f6-ce53b0620367.png" width="550">
<img src="https://user-images.githubusercontent.com/16495490/59074320-e7fc7800-8898-11e9-8303-8926c91150be.png" width="400">
<img src="https://user-images.githubusercontent.com/16495490/59075970-812f8c80-88a1-11e9-96c7-ce14529b7adb.png" width="500">

  [1]: https://arxiv.org/abs/1703.04046
  [2]: https://github.com/titu1994/keras-squeeze-excite-network/blob/master/se.pyom/BXtHS.png

**Comment by OverLordGoldDragon** [comment]
BatchNormalization outputs with a more dramatic val vs. train difference:
<img src="https://user-images.githubusercontent.com/16495490/59150806-4b9cb780-89f7-11e9-96ae-418a2eacb86f.png" width="500">

**Comment by OverLordGoldDragon** [comment]
_Probably_ resolved: I've inadvertently left in one non-standardized sample (in a batch of 32), with `sigma=52` - which severely disrupted the `BN` layers; post-standardizing, I no longer observe a strong discrepancy between train & inference modes - if anything, any differences are difficult to spot.
